File: frontend/src/types/storage.ts
// storage domain types. ui layer uses camelCase; wire types
// mirror the api response shape (already camelCase after the
// api-client key conversion) and are suffixed with *Wire and
// mapped to domain shapes at the hook boundary.

export type StorageAdvisory = 'proceed' | 'warning' | 'blocked';

export type RelocationStatus = 'running' | 'completed' | 'failed';

export interface StorageUsage {
  dataDir: string;
  freeBytes: number;
  totalBytes: number;
  originalsBytes: number;
  derivativesBytes: number;
  dbBytes: number;
  logsBytes: number;
  assetCount: number;
  onSystemDrive: boolean;
}

export interface StorageUsageWire {
  dataDir: string;
  freeBytes: number;
  totalBytes: number;
  originalsBytes: number;
  derivativesBytes: number;
  dbBytes: number;
  logsBytes: number;
  assetCount: number;
  onSystemDrive: boolean;
}

export interface StorageCheckRequest {
  path: string;
  estimatedBatchSize: number;
}

export interface StorageCheckResult {
  writable: boolean;
  writableReason: string | null;
  freeBytes: number;
  totalBytes: number;
  onSystemDrive: boolean;
  advisory: StorageAdvisory;
  advisoryReason: string | null;
}

export interface StorageCheckResultWire {
  writable: boolean;
  writableReason: string | null;
  freeBytes: number;
  totalBytes: number;
  onSystemDrive: boolean;
  advisory: StorageAdvisory;
  advisoryReason: string | null;
}

export interface RelocateRequest {
  targetPath: string;
}

export interface RelocateAcceptedResponseWire {
  jobId: string;
}

export interface RelocationJob {
  jobId: string;
  status: RelocationStatus;
  assetsCopied: number;
  assetsTotal: number;
  bytesCopied: number;
  bytesTotal: number;
  error: string | null;
  startedAt: string | null;
  completedAt: string | null;
}

export interface RelocationJobWire {
  jobId: string;
  status: RelocationStatus;
  assetsCopied: number;
  assetsTotal: number;
  bytesCopied: number;
  bytesTotal: number;
  error: string | null;
  startedAt: string | null;
  completedAt: string | null;
}

File: backend/src/loom/schemas/storage.py
"""schemas for the lite-profile storage management endpoints.

exposes usage stats for Settings -> Storage, a pre-ingest free-space
check for the batch-picker flow, and the relocation job envelope
used by the data-directory move wizard. see issue #47.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class StorageUsage(BaseModel):
    """snapshot of the lite-profile data directory footprint."""

    data_dir: str
    free_bytes: int
    total_bytes: int
    originals_bytes: int
    derivatives_bytes: int
    db_bytes: int
    logs_bytes: int
    asset_count: int
    # true when the data dir lives on the same drive as the OS —
    # triggers a "move to external" advisory for field installs.
    on_system_drive: bool


class StorageCheckRequest(BaseModel):
    """ask whether ``path`` has room for a planned ingest batch."""

    path: str
    estimated_batch_size: int = Field(ge=0)


class StorageCheckResponse(BaseModel):
    writable: bool
    writable_reason: str | None = None
    free_bytes: int
    total_bytes: int
    on_system_drive: bool
    advisory: Literal["proceed", "warning", "blocked"]
    advisory_reason: str | None = None


class RelocationRequest(BaseModel):
    target_path: str


class RelocationJobStatus(BaseModel):
    job_id: str
    status: Literal["running", "completed", "failed"]
    assets_copied: int
    assets_total: int
    bytes_copied: int
    bytes_total: int
    error: str | None = None
    started_at: datetime
    completed_at: datetime | None = None


class RelocationStartResponse(BaseModel):
    job_id: str
